consensus/ibft/fork/manager.go

```diff
@@ -0,0 +1,424 @@
+	validatorSets map[store.SourceType]store.ValidatorStore
+}
+
+// NewForkManager is a constructor of ForkManager
+func NewForkManager(
+	logger hclog.Logger,
+	blockchain store.HeaderGetter,
+	executor contract.Executor,
+	secretManager secrets.SecretsManager,
+	filePath string,
+	epochSize uint64,
+	ibftConfig map[string]interface{},
+) (*ForkManager, error) {
+	forks, err := GetIBFTForks(ibftConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	fm := &ForkManager{
+		logger:         logger.Named(loggerName),
+		blockchain:     blockchain,
+		executor:       executor,
+		secretsManager: secretManager,
+		filePath:       filePath,
+		epochSize:      epochSize,
+		forks:          forks,
+		signers:        make(map[validators.ValidatorType]signer.Signer),
+		validatorSets:  make(map[store.SourceType]store.ValidatorStore),
+	}
+
+	// Need initialization of signers in the constructor
+	// because hash calculation is called from blockchain initialization
+	if err := fm.initializeSigners(); err != nil {
+		return nil, err
+	}
+
+	return fm, nil
+}
+
+// Initialize initializes ForkManager on initialization phase
+func (m *ForkManager) Initialize() error {
+	if err := m.initializeValidatorStores(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetSigner returns a proper signer at specified height
+func (m *ForkManager) GetSigner(height uint64) (signer.Signer, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	signer, ok := m.signers[fork.ValidatorType]
+	if !ok {
+		return nil, ErrSignerNotFound
+	}
+
+	return signer, nil
+}
+
+// GetValidatorStore returns a proper validator set at specified height
+func (m *ForkManager) GetValidatorStore(height uint64) (store.ValidatorStore, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	set, ok := m.validatorSets[ibftTypesToSourceType[fork.Type]]
+	if !ok {
+		return nil, ErrValidatorStoreNotFound
+	}
+
+	return set, nil
+}
+
+// GetValidators returns validators at specified height
+func (m *ForkManager) GetValidators(height uint64) (validators.Validators, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	set, err := m.GetValidatorStore(height)
+	if err != nil {
+		return nil, err
+	}
+
+	return set.GetValidators(
+		calculateFetchingValidatorsHeight(height, m.epochSize, fork),
+	)
+}
+
+// GetHooks returns a hooks at specified height
+func (m *ForkManager) GetHooks(height uint64) (*hook.Hooks, error) {
+	hooks := &hook.Hooks{}
+
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	var err error
+
+	switch fork.Type {
+	case PoA:
+		err = m.registerPoAHooks(hooks, height)
+	case PoS:
+		registerPoSHook(hooks, m.epochSize)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := m.registerPoAPrepareHooks(hooks, height); err != nil {
+		return nil, err
+	}
+
+	m.registerPoSPrepareHooks(hooks, height)
```

Comment:
How can we have both hook types registered here ?

Reply:
Could you tell me more details?

Reply:
I figured PoS and PoA hooks are entirely disjoint sets of methods. From this piece, it looks like we can have both (at the same time).